src/transactions/TransactionFrame.cpp

```diff
@@ -471,57 +489,64 @@ TransactionFrame::processFeeSeqNum(AbstractLedgerTxn& ltx, int64_t baseFee)
 {
-    for (auto const& usedAccount : signatureChecker.usedOneTimeSignerKeys())
+    auto header = ltx.loadHeader();
+    auto ledgerVersion = header.current().ledgerVersion;
+
+    if (ledgerVersion == 7)
     {
-        removeUsedOneTimeSignerKeys(ltx, usedAccount.first, usedAccount.second);
+        return;
     }
-}
 
-void
-TransactionFrame::removeUsedOneTimeSignerKeys(
-    AbstractLedgerTxn& ltx, AccountID const& accountID,
-    std::set<SignerKey> const& keys) const
-{
-    auto account = stellar::loadAccount(ltx, accountID);
-    if (!account)
+    std::unordered_set<AccountID> accounts{getSourceID()};
+    for (auto& op : mOperations)
     {
-        return; // probably account was removed due to merge operation
+        accounts.emplace(op->getSourceID());
     }
 
-    auto header = ltx.loadHeader();
-    auto changed = std::accumulate(
-        std::begin(keys), std::end(keys), false,
-        [&](bool r, const SignerKey& signerKey) {
-            return r || removeAccountSigner(header, account, signerKey);
-        });
-
-    if (changed)
+    auto signerKey = SignerKeyUtils::preAuthTxKey(*this);
+    for (auto const& accountID : accounts)
     {
-        normalizeSigners(account);
+        removeAccountSigner(ltx, header, accountID, signerKey);
     }
 }
 
-bool
-TransactionFrame::removeAccountSigner(LedgerTxnHeader const& header,
-                                      LedgerTxnEntry& account,
+void
+TransactionFrame::removeAccountSigner(AbstractLedgerTxn& ltx,
+                                      LedgerTxnHeader const& header,
+                                      AccountID const& accountID,
                                       SignerKey const& signerKey) const
 {
-    auto& acc = account.current().data.account();
+    auto accountNoRecord = stellar::loadAccountWithoutRecord(ltx, accountID);
+    if (!accountNoRecord)
+    {
+        // account was probably removed due to merge operation
+        return;
+    }
+
+    auto& signersNoRecord = accountNoRecord.current().data.account().signers;
     auto it = std::find_if(
-        std::begin(acc.signers), std::end(acc.signers),
+        std::begin(signersNoRecord), std::end(signersNoRecord),
         [&signerKey](Signer const& signer) { return signer.key == signerKey; });
-    if (it != std::end(acc.signers))
+
+    if (it != std::end(signersNoRecord))
     {
+        auto indexToRemove = it - std::begin(signersNoRecord);
+        auto keyToRemove = it->key;
+
+        // The signer exists, so we can load the account to modify it now
+        accountNoRecord.deactivate();
+        auto account = stellar::loadAccount(ltx, accountID);
+
         auto removed = stellar::addNumEntries(header, account, -1);
         assert(removed == AddSubentryResult::SUCCESS);
-        acc.signers.erase(it);
-        return true;
-    }
 
-    return false;
+        auto& signers = account.current().data.account().signers;
+        assert(signers[indexToRemove].key == keyToRemove);
```

Comment:
For extra safety, you should `assert` that `indexToRemove < size`. I've done something similar in CAP-0015, where I'm using the same approach in the `TransactionQueue`: https://github.com/stellar/stellar-core/blob/36c59b503d5c0e3809063e763871a1b501fd4e54/src/herder/TransactionQueue.cpp#L80-L81